Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/FileSystem.h

cdumez opened the pull request:
#### 1bdae87849777fb04762fdb0169b93ade8019a71
<pre>
Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/FileSystem.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=282801">https://bugs.webkit.org/show_bug.cgi?id=282801</a>

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::MappedFileData::leakHandle):
(WTF::FileSystemImpl::MappedFileData::operator bool const):
(WTF::FileSystemImpl::MappedFileData::size const):
(WTF::FileSystemImpl::MappedFileData::span const):
(WTF::FileSystemImpl::MappedFileData::mutableSpan):
(WTF::FileSystemImpl::MappedFileData::fileMapping const):
* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::~MallocSpan):
(WTF::MallocSpan::mmap):
* Source/WTF/wtf/Mmap.h: Added.
(WTF::Mmap::mmap):
(WTF::Mmap::free):
* Source/WTF/wtf/TypeTraits.h:
(WTF::parameterCount):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::allocateBuffer):
(IPC::Encoder::reserve):
(IPC::Encoder::messageFlags):
(IPC::Encoder::messageFlags const):
(IPC::Encoder::grow):
(IPC::Encoder::capacityBuffer):
(IPC::Encoder::capacityBuffer const):
(IPC::allocBuffer): Deleted.
(IPC::freeBuffer): Deleted.
(IPC::Encoder::~Encoder): Deleted.
(IPC::Encoder::freeBufferIfNecessary): Deleted.
* Source/WebKit/Platform/IPC/Encoder.h:

Canonical link: <a href="https://commits.webkit.org/286374@main">https://commits.webkit.org/286374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d162e9e230a9226c04ba482214ea5c562cdff49d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75813 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80304 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27079 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59449 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17613 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49329 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65118 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39808 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22601 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25406 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68990 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81774 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75089 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3182 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1998 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67686 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66991 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10927 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9057 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97343 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11712 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3132 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5937 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21281 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3153 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4091 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -365,7 +363,7 @@ bool MappedFileData::mapFileHandle(PlatformFileHandle handle, FileOpenMode openM
         return false;
 
-    m_fileData = { static_cast<uint8_t*>(data), size };
+    m_fileData = unsafeMakeSpan(static_cast<uint8_t*>(data), size);
```
It would be even better to use `MallocSpan` here to guarantee `munmap` gets called.

cdumez replied:
There are issues with reusing `MallocSpan` for this:
- Malloc::free() currently only take a pointer, no size (Malloc being the template parameter)
- Malloc::malloc() only takes a size but `mmap()` would take a lot more parameters.

I agree with should have some sort of RAII object to avoid memory leaks. However, I'm not sure reusing MallocSpan is suitable here. What do you think?

darinadler replied:
I don’t really care much. `MappedFileData` itself is a sort of RAII object, just has too much other code. So maybe we just need to refactor into two classes to separate concerns better. Seems like in theory a template like `MallocSpan` could be designed to be flexible about the arguments to allocation and include the size for the free function, since the size is what makes a span. But I understand that’s not how the actual template is designed, and that’s OK.

darinadler commented on Source/WebKit/Platform/IPC/Encoder.cpp:
```diff
@@ -72,17 +65,8 @@ Encoder::Encoder(MessageName messageName, uint64_t destinationID)
-    if (m_capacityBuffer.data() != m_inlineBuffer.data())
-        freeBuffer(m_capacityBuffer);
-}
+// FIXME: We need to dispose of the attachments in cases of failure.
```
Good catch! I guess we need to create that test case too.

cdumez replied:
I merely moved this comment.

darinadler replied:
Nerve wracking!

cdumez committed:
Drop use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/FileSystem.h
https://bugs.webkit.org/show_bug.cgi?id=282801

Reviewed by Darin Adler.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::mapFileHandle):
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::MappedFileData::leakHandle):
(WTF::FileSystemImpl::MappedFileData::operator bool const):
(WTF::FileSystemImpl::MappedFileData::size const):
(WTF::FileSystemImpl::MappedFileData::span const):
(WTF::FileSystemImpl::MappedFileData::mutableSpan):
(WTF::FileSystemImpl::MappedFileData::fileMapping const):
* Source/WTF/wtf/MallocSpan.h:
(WTF::MallocSpan::~MallocSpan):
(WTF::MallocSpan::mmap):
* Source/WTF/wtf/Mmap.h: Added.
(WTF::Mmap::mmap):
(WTF::Mmap::free):
* Source/WTF/wtf/TypeTraits.h:
(WTF::parameterCount):
* Source/WebKit/Platform/IPC/Encoder.cpp:
(IPC::allocateBuffer):
(IPC::Encoder::reserve):
(IPC::Encoder::messageFlags):
(IPC::Encoder::messageFlags const):
(IPC::Encoder::grow):
(IPC::Encoder::capacityBuffer):
(IPC::Encoder::capacityBuffer const):
(IPC::allocBuffer): Deleted.
(IPC::freeBuffer): Deleted.
(IPC::Encoder::~Encoder): Deleted.
(IPC::Encoder::freeBufferIfNecessary): Deleted.
* Source/WebKit/Platform/IPC/Encoder.h:

Canonical link: https://commits.webkit.org/286374@main
